man/dbdind.1
.TH DBDIND 1 "12 September 1995"
.FB
.SH NAME
dbdind \- define a \fIFirstBase\fP index generator dictionary
.SH SYNOPSIS
.B dbdind
[\fB-d\fP \fIdatabase\fP] [\fB-i\fP \fIindex\fP] [\fB-n\fP]
.SH DESCRIPTION
.I Dbdind
allows definition and modification of a 
\fIFirstBase\fP index generator dictionary via a screen oriented editor.
This dictionary is then used by
\fIdbigen\fP(1)
to produce a \fIFirstBase\fP 
.I index.
.PP
Note: the default index produced is a FirstBase Btree+ index.
.PP
Many tools in the \fIFirstBase\fP use indexes to provide structure and
order for a database. Any index created can be used anywhere an index
is allowed \(em regardless of the size and shape of the index.
An index can contain an entry for each record of a database, or a selected
few. \fIDbdind\fP(1) is the formal mechanism for defining the index dictionary
needed by \fIdbigen\fP(1) to create the index.
.PP
The index dictionary is a list of selection criteria items
that are used as a road map to determine
which records of a database are to be referenced by the index-to-be.
This list of items is held together by 'logical connectors' \(em 
connectors that allow a selective road map to be built.
.PP
Often, the index dictionary contains a single criterion
used to select all records of a database.
This complete index is used merely
to impose some form of sorted order on a database.
.PP
Other times, the index dictionary contains many selection criteria.
In these cases,
a record from the database is included in the final index only if
it passes some or all of the tests indicated by each
selection criterion.
.PP
A single selection criterion
consists of a valid database field name along with
two values that govern selection of records based on the contents of
the specified field within each record.
These two selection values
effectively collect a selected group of records that fall
within specific ranges of values for that
particular database field.
.PP
Two or more selection criteria items are bound together by a logical
connector \(em either an 'AND' or an 'OR' connection. By placing many
selection criteria together and binding them with logical connections,
a very selective index can be created.
.PP
For example, if you wanted a list of all clients from a database that
are between the ages of 25 and 30 AND live in New York AND are
teachers, you would have three selection criteria \(em each bound together with
the logical connector 'AND'.
.PP
In this example, the first criterion is 'Age', and the age must fall
between 25 and 30. The '25' and the '30' are the two selection values
for the Age item that provide an acceptable range of values.
The next selection criterion is 'Home Address', and
the desired range is very small \(em 'New York'. These two selection
criteria \(em 'Age' and 'Home Address' \(em are bound together with
an 'AND' connector. The same follows for the third criterion, 'Profession'.
.PP
Of the possible logical connectors, the 'And' has a higher precedence,
meaning it is stronger than any 'Or' connections. However, there
is a 'Simple And' connector that can be used in the midst of 'Or'
connectors. This 'Simple And' has lower precedence than the 'Or'
connector.
.PP
As mentioned, each selection criterion consists of a field name along with
two selection values. Special 'meta-value' entries are used as these selection
values to limit selection, gather all records, or even to negate
a criterion.
.PP
To capture all the records of a database, regardless of what is stored in
the database under the listed selection field_name is accomplished
by using the meta-value '$ALL' as the first selection value.
This is the default for the first selection value.
.PP
Records that match
one value exactly can be selected by entering the first value and
using the meta-value '$NONE' as the second value. This is the
default for the second selection value.
.PP
Records that have
a null or empty value can be selected by
using the meta-value '$EMPTY' as the first value. Alternately, '$NOT' can
be the first value, '$EMPTY' the second value to capture records that
DO NOT have an empty or null value.
.PP
Records that match
a UNIX style regular expression
can be selected by entering the regular expression as the first value,
and using the meta-value '$PATTERN' as the second value.
(Use a '\\' to escape an underscore since underscores usually represent
blanks).
.PP
NOTE: Under BSD UNIX, multiple regular expressions in a single dictionary
are expensive, but they do work.
Using SYS V UNIX, multiple patterns are easily handled.
.PP
Selection of records with
field entries NOT equal to a value can be
specified by using the meta-value '$NOT' as the first value, and the value
to exclude as the second value.
For example, this would be used to capture all clients not living in Phoenix.
.PP
A wildcard mechanism for selection of all entries that begin (or do not
begin) with a pattern also exists. To use this feature, put an
asterisk ('*') as the last character of the pattern to match.
.PP
For example, using "Smith*" for a selection value will capture all records
whose key field
begins with 'Smith', without considering 'Smythe' or other values.
This wildcard feature is really only useful when used in conjunction with
meta-values of $NOT, $NONE and $ALL.
.PP
The following is a list of all index dictionary attributes
and their purpose:
.TP 7
.B field
is the actual name of the already-defined field from
the database dictionary. This entry defaults to the first field in the
database dictionary.
.TP 7
.B Val1
is the greater than or equal value point, i.e. the left end of the
range of records to trap. The default for the entry is $ALL, meaning
all records that have ANY data in this field will be chosen.
.sp 1
If
.B $NOT
is used, then records will pass this test only if their field value is 
.B not equal
to the value indicated in the Val2 field of this selection element.
.sp 1

ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile


src/dbigen/dbdind.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbdind.c,v 9.1 2001/02/16 19:43:51 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char sccsid[] = "@(#)dbdind.c	8.3 05 Apr 1996 FB";
#endif

/* 
 *  dbdind.c - screen editor for index generator dictionary file: idicti
 *     this program is unnecessarilly complicated due to lack of forethought.
 *     the problem is that i used the same structure to edit as i do for
 *     actually generating indexes from a selection tree.
 *     tsk tsk. deep sigh...(this works though!, and quite well.)
 */

#include <fb.h>
#include <fb_vars.h>
#include <igen.h>
#include <dbd.h>

int btree = 1;			/* gen a Btree index when done */

#define MAXWORD 20
#define SHOWP 4			/* zero based limit of screen postioins */

#define HLP_DBDIND 	"dbdind.hlp"
#define DBDIND_VAL1	"dbdind1.hlp"
#define DBDIND_VAL2	"dbdind2.hlp"
#define DBDIND_VAL3	"dbdind3.hlp"

int col[8] = {1, 6, 28, 50, 72};

int dot = 0,			/* current pointer...postional from top */
    last = 0;			/* last andtree...postional from top */

fb_field *by[FB_MAXBY];		/* sortby fields */
struct self *top = NULL;	/* top pointer */

static fb_database *hp;
int cflag = 0;			/* case-insensative flag */

extern short int cdb_locklevel;
extern short int cdb_askgen;
extern char *cdb_IGEN;
extern short int cdb_choosefield;
extern char *cdb_T_ALL;
extern char *cdb_T_NOT;
extern char *cdb_T_NONE;
extern char *cdb_T_NOTEQUAL;
extern char *cdb_T_EQUAL;
extern char *cdb_T_LESSEQUAL;
extern char *cdb_T_GREATEREQUAL;
extern char *cdb_T_LESSTHAN;
extern char *cdb_T_GREATERTHAN;
extern char *cdb_T_PATTERN;
extern char *cdb_T_EMPTY;

static no_chooseval = 0;

static struct self *bpos();	/* function to get pointer into self list */
static iedit();
static fillscreen();
static escreen();
static void fdelete();
static fadder();
static efield();
static fput();
static output();
static convert();
static headers();
static begin();
static anychange();

/*
 *  dbdind - main driver.
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
      
      {
         struct self *stree();
         int i, st;

         cdb_locklevel = -1;
         fb_getargs(argc, argv, FB_ALLOCDB);
	 hp = cdb_db;
	 fb_initdbd(hp);
         if (fb_getd_dict(hp) == FB_ERROR)
            fb_xerror(FB_BAD_DICT, hp->ddict, NIL);
	 if (fb_testargs(argc, argv, "-n") > 0)
	    no_chooseval = 1;
	 begin();
	 for (i = 0; i < hp->nfields; i++)
	    fb_nounders(cdb_keymap[i]);
         if ((top = stree(cdb_keymap, hp, by)) == NULL){
	    for (i = 0; i < FB_MAXBY; i++)
	       by[i] = NULL;
	    }
	 else
	    convert();
	 if ((st = iedit(by)) == FB_END){
	    if (by[0] == NULL){
	       getby(by);
	       if (cdb_askgen)
	          headers();
	       }
	    }
	 if (st == FB_END){
	    output(by);
	    if (cdb_askgen)
	       fb_chainto("Generate Index", cdb_IGEN, argv);
	    }
	 else
	    fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
         fb_ender();
      }

/* 
 * iedit - edit the self structures 
 */
 
   static iedit(by)
      fb_field *by[];
   
      {
         int st, i;
	 struct self *p, *q, *r;

         dot = 1;
	 p = top;
	 for (i = 0; p != NULL; ){
	    for(q = p, r = p; q != NULL; q = q->orp){
	       i++;
	       r = q;
	       }
	    if (r == p)
	       p = p->andp;
	    else
	       p = r->andp;
	    }
	 last = i;
         fillscreen();
         while ((st = escreen(by)) != FB_END){
            if (st == FB_ABORT){
	       if (fb_mustbe('y', "Really Quit? (y/n) ", cdb_t_lines, 1) ==
                     FB_AOK)
		  break;
	       }
            else if (st >= dot && st <= dot + 9 && st <= last)
               efield(st, 0, (struct self *) NULL);
            else if (st >= 1 && st <= last){
               dot = fb_onpage(st);
               fillscreen();
               }
            else
               fb_screrr("Invalid Input");
            }
         return(st);
      }

/* 
 * fillscreen - fill a screen with fields 
 */
 
   static fillscreen()
      {
	 char type;
         int j, i, k;
	 struct self *q, *r;

         fb_move(4,1); fb_clrtobot();
	 r = bpos(dot);
	 if (r == NULL){
	    q = top;
	    r = top;
	    dot = 1;
	    }
	 else 
	    q = r->orp;
         j = dot + 9;
         for (i = dot; i <= j && i <= last; i++, r = q, q = q->orp){
	    if (r->orp != NULL && r->orp == r->sandp)
	       type = 's';
	    else if (q == NULL){
	       if ((r == NULL) || ((q = r->andp) == NULL))
	          break;
	       type = 'a';			/* and = 'a' , or = 'o' */
	       }
	    else
	       type = 'o';
	    for (k = 0; k <= SHOWP; ++k)
	       fput(q, i, k, type);
	    }
	 fb_infoline();
         fb_cx_set_viewpage(dot);
      }

/* 
 * escreen - set up a screen, accept commands for screen control 
 */
 
   static escreen(by)
      fb_field *by[];
      
      {
         int st;
	 char com[5], buf[FB_MAXLINE];
         
         for (;;){
            fb_cx_push_env("E1", CX_DBD_SELECT, NIL);
            if (last > 10)
               fb_cx_add_buttons("Ge");
            fb_cx_add_buttons("sXH");
            if (btree)
               strcpy(buf, "BTree+ Index");
            else
               strcpy(buf, "Flat Index");
            st = anychange(com, buf);
            fb_cx_pop_env();
            switch (st) {
               case FB_ABORT: case FB_END:
                  return(st);
               case ADDMODE:
                  fadder(); 
		  break;
               case DELMODE:
                  fdelete(); 
		  break;
	       case '?':				/* FB_HELP */
	          fb_move(2, 1);
		  fb_clrtobot();
	          fb_help(com, hp);
		  headers();
		  break;
               case PAGEF:
                  dot = dot + 10;
                  if (dot > last)
                     dot = 1;
                  break;
               case PAGEB:
                  dot = dot - 10;
                  if (dot < 1)
                     dot = fb_onpage(last);
                  break;
	       case FB_QHELP:
	          fb_move(2, 1); fb_clrtobot();
	          fb_fhelp(HLP_DBDIND);
		  headers();
		  break;
	       case SORTBYMODE:
	          getby(by);
		  headers();
		  break;
	       case TRACEMODE:
	          fb_move(2, 1); fb_clrtobot();
	          itrace(top, by, 1);
		  headers();
		  break;
	       case FB_AOK:
                  if (equal(com, "tree") || equal(com, "norm")){
                     if (btree)
                        btree = 0;
                     else
                        btree = 1;
                     continue;
                     }
                  return(atoi(com));
               default:
                  return(st);
               }
            fillscreen();
            }
      }

/* 
 * fdelete - delete a fb_field 
 */
 
   static void fdelete()
      {
         int st, n;
	 struct self *p, *q;
         
         fb_fmessage("Delete what number ? (#,-) ");
         st = fb_input(cdb_t_lines, 28, 3, 0, FB_INTEGER, (char *) &n,
            FB_ECHO,FB_OKEND,FB_CONFIRM);
         if (st == FB_END || st == FB_ABORT)
            return;
	 else if (n < 1 || n > last)
	    return;
	 if ((p = bpos(n)) == NULL){
            p = top;
	    if ((top = p->orp) == NULL)
	       top = p->andp;
	    fb_free((char *) p);
	    }
	 else{
	    if ((q = p->andp) == NULL)
	       q = p->orp;
	    p->andp = NULL;
	    p->orp = NULL;
	    if (q != NULL && q->andp != NULL)
	       p->andp = q->andp;
	    else if (q != NULL && q->orp != NULL)
	       p->orp = q->orp;
	    fb_free((char *) q);
	    }
         last--;
      }

/* 
 * fadder - go into auto add mode 
 */
 
   static fadder()
      {
         struct self *p, *q;
	 
         dot = fb_onpage(last);
	 if ((last + 1) <  (dot + 10))
            fillscreen();
         for (;;){
            p = (struct self *) fb_malloc(sizeof(struct self));
	    p->andp = NULL;
	    p->orp = NULL;
	    p->lword[0] = NULL;
	    p->rword[0] = NULL;
            if (++last >= dot + 10){
               dot += 10;
               fb_move(4, 1); fb_clrtobot(), fb_infoline();
               }
            if (efield(last, 1, p) == FB_END){
	       q = bpos(last);
	       if (q == NULL)
	          q = top;
	       if (q != NULL){
		  q->andp = NULL;
		  q->orp = NULL;
		  }
	       if (top == p)
	          top = NULL;
               fb_free((char *) p);
	       last--;
               break;
               }
	    if (cdb_choosefield && !no_chooseval){
	       if (fb_mustbe('y', "Add More Entries (y=Yes, -=End) ?",
                     cdb_t_lines, 1) != FB_AOK)
	          break;
	       }
            }
	 dot = 1;
      }
      
/* 
 * efield - edit an entire index self structure 
 */
 
   static efield(fld, new, newp)
      int fld, new;
      struct self *newp;
      
      {
         int row, i, st, comflag;
         char buf[FB_MAXLINE], *fb_trim(), temp[FB_MAXLINE], msg[FB_MAXLINE];
	 struct self *p, *q;

         strcpy(buf, cdb_T_EMPTY);
         if ((row = (fld-- % 10)) == 0)
            row = 10;
         row = row * 2 + 2;
         fb_move(cdb_t_lines, 1); fb_clrtoeol();
	 q = bpos(fld+1);
	 p = NULL;
         if (new == 1){
	    if (q != NULL){
	       q->andp = newp;
	       q->orp = newp;
	       q->sandp = NULL;
	       }
            strcpy(msg, "Enter Field, <RET>=Default, -=End");
            fput(p, fld+1, 0, FB_BLANK);	/* puts number on screen */
	    if (top == NULL)
	       top = newp;
            }
         else
            strcpy(msg, "Change Field, <CTL>-X=Skip, <RET>=Default");
	 if (q == NULL)
	    p = top;
	 else if (q->orp == NULL)
	    p = q->andp;
	 else 
	    p = q->orp;
	 
	 /*
	  *  get fp (fb_field pointer)...via fieldname
	  */

         for (;;){
	    if (!cdb_choosefield){
	       fb_fmessage(msg);
	       st = fb_input(row, col[1], FB_TITLESIZE, 0, FB_ALPHA, 
		    temp, FB_ECHO, FB_OKEND, FB_CONFIRM);
	       }
	    else{
	       st = fb_choosefield(temp);
	       headers();
	       }
            if (st == FB_END && new == 1){
               fb_move(row, 1); 
               fb_clrtoeol();
               return(FB_END);
               }
            else if (st == FB_END || (st == FB_ABORT && new == 0))
               break;
            if (st == FB_DEFAULT || (st == FB_ABORT && new == 1)){
	       p->fp = cdb_kp[0];
	       break;
	       }
            else if (st == FB_AOK){
	       fb_trim(temp);
	       for (i = 0; i < hp->nfields; i++)
		  if (strcmp(temp, cdb_kp[i]->id) == 0)
		     break;
	       if (i < hp->nfields){
		  p->fp = cdb_kp[i];
                  if (p->fp->type != FB_BINARY)
		     break;
                  else
	             fb_screrr("Invalid: Cannot Use Binary Field in that Manner");
		  }
               else
	          fb_screrr("Invalid Field: Not in Database Dictionary");
	       }
            }
	 if (cdb_choosefield){
	    if (!new){
	       fillscreen();
	       }
	    else{				/* must be new */
	       last--;
	       fillscreen();
	       last++;
	       fput(p, fld+1, 0, FB_BLANK);	/* puts number on screen */
	       }
	    }
         fput(p, fld+1, 1, FB_BLANK);
	 
	 /* 
	  *  get lword (left word or val1)
	  */

         for(;;){
	    if (!cdb_choosefield || no_chooseval){
	       fb_fmessage(msg);
	       fb_scrhlp("[or use '$ALL', '$NOT', $OP]");
	       st = fb_input(row, col[2], MAXWORD, 0, FB_ALPHA, buf, 
	          FB_ECHO, FB_NOEND, FB_CONFIRM);
	       }
	    else{
	       st = fb_chooseval(DBDIND_VAL1, buf, MAXWORD, 1);
	       headers();
	       }
            if (st == FB_DEFAULT || (st == FB_ABORT && new == 1))
               strcpy(buf, cdb_T_ALL);
	    else if (st == FB_ABORT || st == FB_END)
	       break;
/* 
*	    if (buf[0] == '$' && (buf[1] == 'N' || buf[1] == 'n'))
*	       strcpy(buf, cdb_T_NOT);
*/
	    if (buf[0] == '$' && (buf[1] == 'A' || buf[1] == 'a'))
	       strcpy(buf, cdb_T_ALL);
            strcpy(p->lword, fb_trim(buf));
	    if (equal(p->lword, cdb_T_LESSTHAN) ||
		   equal(p->lword, cdb_T_GREATERTHAN) ||
		   equal(p->lword, cdb_T_LESSEQUAL) ||
		   equal(p->lword, cdb_T_GREATEREQUAL) ||
		   equal(p->lword, cdb_T_NOTEQUAL) ||
		   equal(p->lword, cdb_T_EQUAL))
	       comflag = 1;
	    else
	       comflag = 0;
	    if (st == FB_AOK && !equal(buf, cdb_T_NOT) &&
                  !equal(buf, cdb_T_EMPTY) && 
	          comflag == 0 && !equal(buf, cdb_T_ALL) &&
	          p->fp->type ==FB_DATE && strlen(p->lword) != 6)
	       fb_screrr("Must use 6 characters for DATE selection");
	    else
	       break;
	    }
	 if (cdb_choosefield && !no_chooseval){
	    if (!new){
	       fillscreen();
	       }
	    else{				/* must be new */
	       last--;
	       fillscreen();
	       last++;
	       fput(p, fld+1, 0, FB_BLANK);	/* puts number on screen */
	       fput(p, fld+1, 1, FB_BLANK);	/* puts id on screen */
	       }
	    }
         fput(p, fld+1, 2, FB_BLANK);
	 
	 /*
	  *  get rword (rightword or val2)
	  */

	 if (equal(p->lword, cdb_T_LESSTHAN) ||
		equal(p->lword, cdb_T_GREATERTHAN) ||
		equal(p->lword, cdb_T_LESSEQUAL) ||
		equal(p->lword, cdb_T_GREATEREQUAL) ||
		equal(p->lword, cdb_T_NOTEQUAL) ||
		equal(p->lword, cdb_T_EQUAL))
	    comflag = 1;
	 else
	    comflag = 0;
         if (!equal(cdb_T_ALL, p->lword) && !equal(cdb_T_EMPTY, p->lword) &&
               comflag == 0){
	    /* normal fb_input of rword */
            for(;;){
	       if (!cdb_choosefield || no_chooseval){
		  fb_fmessage(msg);
		  if (!equal(cdb_T_NOT, p->lword))
		     fb_scrhlp("[or use '$NONE' or '$PATTERN']");
		  st = fb_input(row,col[3],MAXWORD,0,FB_ALPHA, buf,
			FB_ECHO, FB_NOEND, FB_CONFIRM);
		  }
	       else{
		  st = fb_chooseval(DBDIND_VAL2, buf, MAXWORD, 2);
		  headers();
		  }
	       fb_trim(buf);
               if (st == FB_DEFAULT || (st == FB_ABORT && new == 1))
                  strcpy(buf, cdb_T_NONE);
	       else if (st == FB_ABORT || st == FB_END)
	          break;
	       else if (buf[0] == '$' && (buf[1] == 'N' || buf[1] == 'n'))
	          strcpy(buf, cdb_T_NONE);
	       else if (buf[0] == '$' && (buf[1] == 'P' || buf[1] == 'p'))
	          strcpy(buf, cdb_T_PATTERN);
	       if (strcmp(p->lword, cdb_T_NOT) == 0 && 
	             (st == FB_DEFAULT || equal(buf, cdb_T_PATTERN))){
	          fb_screrr("Must enter value here: Val1 is $NOT!");
		  continue;
		  }
	       if (equal(buf, cdb_T_EMPTY) && !equal(p->lword, cdb_T_NOT)){
	          fb_screrr("Must use $NOT as Val1 when $EMPTY is Val2");
		  continue;
                  }
               strcpy(p->rword, buf);
	       if (st == FB_AOK && p->fp->type == FB_DATE && strlen(p->rword) != 6 &&
                  !equal(p->rword, cdb_T_EMPTY))
		  fb_screrr("Must use 6 characters for DATE selection");
	       else
		  break;
               }
	    }
	 else if (!equal(cdb_T_ALL, p->lword) && !equal(cdb_T_EMPTY, p->lword) &&
               comflag == 1){
	    /* selection of a compare fb_field -- lword was a connective OP */
	    for (;;){
	       if (!cdb_choosefield){
		  fb_fmessage(msg);
		  st = fb_input(row, col[3], FB_TITLESIZE, 0, FB_ALPHA, 
		       temp, FB_ECHO, FB_NOEND, FB_CONFIRM);
		  }
	       else{
		  st = fb_choosefield(temp);
		  headers();
		  }
	       if ((st == FB_ABORT || st == FB_END) && new == 0){
	          strcpy(temp, p->rword);
		  st = FB_AOK;
		  }
	       if (st == FB_DEFAULT || st == FB_END || (st == FB_ABORT && new == 1)){
	          strcpy(temp, cdb_kp[0]->id);
		  st = FB_AOK;
		  }
	       if (st == FB_AOK){
		  fb_trim(temp);
	          if ((p->cfp = fb_findfield(temp, hp)) != NULL){
		     strcpy(p->rword, p->cfp->id);
		     break;
		     }
		  fb_screrr("Invalid Field: Not in Database Dictionary");
		  }
	       }
            }
	 else
	    p->rword[0] = NULL;
	 if (cdb_choosefield){
	    if (!new){
	       fillscreen();
	       }
	    else{				/* must be new */
	       last--;
	       fillscreen();
	       last++;
	       fput(p, fld+1, 0, FB_BLANK);	/* puts number on screen */
	       fput(p, fld+1, 1, FB_BLANK);	/* puts id on screen */
	       fput(p, fld+1, 2, FB_BLANK);	/* puts lword on screen */
	       }
	    }
         fput(p, fld+1, 3, FB_BLANK);
	 
	 /*
	  *  get andp/orp/sanp connective (logical connector)
	  */

         buf[0] = NULL;
         for(;;){
	    if (fld <= 0 || q == NULL)
	       break;
	    if ((fld+1) % 10 == 1)	/* 11, 21, 31, 41... */
	       i = row;
	    else
	       i = row - 1;
	    fb_move(i, col[4]);
	    fb_printw("       ");
	    if (!cdb_choosefield || no_chooseval){
	       fb_fmessage(msg);
	       fb_scrhlp("a=And, o=OR, s=Simple And");
	       st = fb_input(i, col[4], 3, 0, FB_ALPHA, buf, 
		     FB_ECHO, FB_NOEND, FB_CONFIRM);
	       }
	    else{
	       st = fb_chooseval(DBDIND_VAL3, buf, MAXWORD, 3);
	       headers();
	       }
            if (st == FB_DEFAULT || (st == FB_ABORT && new == 1))
               buf[0] = 'a';
            else if (st == FB_ABORT){
	       if (q->andp == p) 
	          buf[0] = 'a';
	       else if (q->orp != NULL && q->orp == q->sandp)
	          buf[0] = 's';
	       else
	          buf[0] = 'o';
               break;
	       }
            if (buf[0] != 'a' && buf[0] != 'o' && buf[0] != 's')
               fb_screrr("a=And, o=OR, s=Simple And");
            else{	/* fix up links to reflect AND or OR selection */
	       if (new == 1){
		  switch(buf[0]) {
		     case 'a': q->orp = NULL; break;
		     case 'o': q->andp = NULL; break;
		     case 's':
		        q->andp = NULL; 
			q->sandp = q->orp;
			break;
		     }
		  }
	       else {
		  switch(buf[0]) {
		     case 'a': 
		        q->sandp = q->orp = NULL;
			q->andp = p;
			break;
		     case 'o':
		        q->sandp = q->andp = NULL;
			q->orp = p;
			break;
		     case 's':
		        q->andp = NULL;
			q->orp = q->sandp = p;
			break;
		     }
		  }
               break;
               }
            }
	 if (cdb_choosefield && !no_chooseval){
	    if (!new){
	       fillscreen();
	       }
	    else{				/* must be new */
	       last--;
	       fillscreen();
	       last++;
	       fput(p, fld+1, 0, FB_BLANK);	/* puts number on screen */
	       fput(p, fld+1, 1, FB_BLANK);	/* puts id on screen */
	       fput(p, fld+1, 2, FB_BLANK);	/* puts lword on screen */
	       fput(p, fld+1, 3, FB_BLANK);	/* puts rword on screen */
	       }
	    }
	 fput(p, fld+1, 4, buf[0]);
	 
	 /*
	  *  finally return
	  */
	  
         return(FB_AOK);
      }

/* 
 *  fput - fb_put an entire fb_field to a constant screen location 
 */
 
   static fput(p, i, pos, aflag)
      struct self *p;
      int i, pos;
      char aflag;
      
      {
         int row, j;
         
         if ((row = (i % 10)) == 0)
            row = 10;
         row = row * 2 + 2;
         fb_move(row, col[pos]);
         switch (pos){
            case 0: fb_printw("%3d)", i); break;
            case 1: fb_printw("%-10s", (p->fp)->id); break;
            case 2: fb_printw("%-20s", p->lword); break;
            case 3: fb_printw("%-20s", p->rword); break;
            case 4: 
	       if (i > 1){
	          if (i % 10 == 1)	/* 11, 21, 31, 41... */
	             j = row;
		  else
		     j = row - 1;
		  fb_move(j, col[pos]);
		  switch (aflag){
		     case 'a':   fb_printw("$AND"); break;
		     case 'o':   fb_printw("$OR"); break;
		     case 's':   fb_printw("$S_AND"); break;
		     }
		  }
	       break;
	    }
      }

/* 
 *  bpos - return postion of self structure before n, NULL if n == top.
 */
 
   static struct self *bpos(n)
      int n;
      
      {
         int i;
	 struct self *p, *r;
	 
	 p = top;	/* traces 'or' trees */
	 r = NULL;	/* to hold last touched good fb_node */
	 for (i = 1; i < n && i < last && p != NULL; i++){
	    r = p;
	    if (p->orp == NULL)
	       p = p->andp;
	    else
	       p = p->orp;
	    }
	 return(r);
      }
      
/* 
 * output - output the tree structure 
 */
 
   static output(by)
      fb_field *by[];
      
      {
         char filen[FB_MAXNAME], out[FB_MAXLINE];
         int i;
	 FILE *fs, *fb_mustfopen();
	 struct self *p, *t, *r = NULL;

	 strcpy(filen, hp->idict);
	 strcat(filen, "i");
	 fs = fb_mustfopen(filen, "w");
	 if (top == NULL){
            top = (struct self *) fb_malloc(sizeof(struct self));
	    top->andp = NULL;
	    top->orp = NULL;
	    top->fp = cdb_kp[0];
	    strcpy(top->lword, cdb_T_ALL);
	    top->rword[0] = NULL;
	    sprintf(out, "No Selection Tree: '%s $ALL' assumed", cdb_kp[0]->id);
	    fb_screrr(out);
	    }
         for (p = top; p != NULL; p = p->andp){
            fprintf(fs, "$\n");
            if (p->orp != NULL){
               fprintf(fs, "#\n");
               for (t = p; t != NULL; r = t, t = t->orp){
	          fb_unders(t->fp);
		  fb_underscore(t->lword, 1);
		  fb_underscore(t->rword, 1);
                  fprintf(fs, "%s %s %s\n", (t->fp)->id, t->lword, t->rword);
		  if (t->orp != NULL && t->orp == t->sandp)
		     fprintf(fs, "&\n");
		  }
               fprintf(fs, "#\n");
	       p = r;
               }
            else{
	       fb_unders(p->fp);
	       fb_underscore(p->lword, 1);
	       fb_underscore(p->rword, 1);
               fprintf(fs, "%s %s %s\n", (p->fp)->id, p->lword, p->rword);
	       }
            }
	 fprintf(fs, "%%\n");
         for (i = 0; i < FB_MAXBY; i++)
            if (by[i] != NULL){
	       fb_unders(by[i]);
               fprintf(fs, "%s ", (by[i])->id);
	       }
	 fprintf(fs, "\n");
         if (btree)
	    fprintf(fs, "%%\n");
	 fclose(fs);
      }

/* 
 * convert - convert dual thread format to single thread format
 */
 
   static convert()
      {
         struct self *p, *q, *r = NULL;
	 
	 for (p = top; p != NULL; p = r->andp){
	    for(q = p; q != NULL; r = q, q = q->orp)
	       ;
	    if (r != p){
	       r->andp = p->andp;
	       p->andp = NULL;
	       }
	    }
      }
   
/* 
 * headers - print header at top of screen 
 */
 
   static headers()
      {
         fb_move(2,col[1]); fb_clrtoeol(); fb_printw("Field");
         fb_move(3,col[1]); fb_clrtoeol(); fb_printw("=====");
         fb_move(2,col[2]); fb_printw("Val1 '>=' or $OP");
         fb_move(3,col[2]); fb_printw("================");
         fb_move(2,col[3]); fb_printw("Val2 '<=' or Field");
         fb_move(3,col[3]); fb_printw("==================");
         fb_move(2,col[4]); fb_printw("And/Or");
         fb_move(3,col[4]); fb_printw("======");
      }

/* 
 * begin  -  refresh the screen template
 */
 
   static begin()
      {
         fb_scrhdr(hp, "Define Index");
	 headers();
      }
      
/* 
 *  anychange - set up anychange screen 
 */
 
   static anychange(inp, msg)
      char inp[], *msg;

      {
         register int st;
         
         fb_fmessage("Any Change (#, -=End, <CTL>-H=HELP) ?");
         fb_scrhlp(msg);
         st = fb_input(-cdb_t_lines, 39, 4, 0, FB_ALPHA, inp, FB_ECHO,
            FB_OKEND, FB_CONFIRM);
	 				/* neg row allows FB_QHELP signal */
         if (st == FB_ABORT || st == FB_END || st == FB_QHELP)
            return(st);
         else if (st == FB_DEFAULT)
            return(PAGEF);
         else if (st == FB_AOK && (equal(inp, "tree") || equal(inp, "norm")))
            return(FB_AOK);
	 switch(inp[0]){
	    case PAGEF:
	    case PAGEB:
	    case ADDMODE:
	    case DELMODE:
	    case INSERTMODE:
	    case HELPMODE:
	    case MINFOMODE:
	    case TRACEMODE:
	    case SORTBYMODE:
	       return(inp[0]);
	    default:
               return(st);
	    }
      }

/* 
 * usage message
 */
 
   usage()
      {
         fb_xerror(FB_MESSAGE, "usage: dbdind [-d dbase] [-i index] [-n]", NIL);
      }
